apps/system/test/marionette/software_home_file_open_error_test.js

```diff
@@ -48,34 +48,69 @@ marionette('Software Home Button - File Open Error', function() {
     system.gotoBrowser(url);
 
     // Save the file.
-    actions.longPress(client.helper.waitForElement('a'), 1).perform();
+    longPress(client.helper.waitForElement('a'));
     client.switchToFrame();
     system.appContextMenuSaveLink.click();
 
     // Tap on the toaster to open the download.
     // We could also open this from settings or the utility tray if needed.
-    var toasterTitle;
-    client.waitFor(function() {
-      toasterTitle = client.helper.waitForElement(
-        '#notification-toaster.displayed .toaster-title');
-      return toasterTitle.text().indexOf('Download complete') !== -1;
-    });
-    toasterTitle.tap();
+    getToasterTitle().tap();
 
-    function rect(el) {
-      return el.getBoundingClientRect();
-    }
+    var dialogHeight = getDialogHeight();
 
-    var dialogHeight = system.downloadDialog.size().height;
     var shbRect = system.softwareButtons.scriptWith(rect);
 
     assert.equal(dialogHeight, expectedDialogHeight());
     assert.equal(dialogHeight, shbRect.top);
   });
 
+  function rect(el) {
+    return el.getBoundingClientRect();
+  }
+
+  function waitForNoException(getter) {
+    var obj;
+    client.waitFor(function() {
+      try {
+        obj = getter();
+        return true;
+      } catch (ex) {
+        return false;
+      }
+    });
+    return obj;
+  }
+
+  function getDialogHeight() {
+    return waitForNoException(function() {
+      return system.downloadDialog.size().height;
+    });
+  }
+
   function expectedDialogHeight() {
     var winHeight = client.findElement('body').size().height;
     var shbHeight = system.softwareButtons.size().height;
     return (winHeight - shbHeight);
   }
+
+  function getToasterTitle() {
+    var toasterTitle;
+    client.waitFor(function() {
+      toasterTitle = client.helper.waitForElement(
+        '#notification-toaster.displayed .toaster-title');
+      return toasterTitle.text().indexOf('Download complete') !== -1;
+    });
+    return toasterTitle;
+  }
+
+  function longPress(el) {
+    var longPressDone = false;
+    actions.longPress(el, 1).perform(function() {
+      longPressDone = true;
+    });
+    client.waitFor(function() {
```

Comment:
Are you sure we need to do this waitFor? It is my understanding that if you don't pass a callback to `perform` it will do a synchronous wait for the action. Can we confirm that?